samples/sensor/ism330dhcx/src/main.c

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (c) 2022 Yestin Sun
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdio.h>
+#include <zephyr.h>
+#include <device.h>
+#include <drivers/sensor.h>
+
+static inline float out_ev(struct sensor_value *val)
+{
+	return (val->val1 + (float)val->val2 / 1000000);
+}
+
+static void fetch_and_display(const struct device *dev)
+{
+	struct sensor_value x, y, z;
+	static int trig_cnt;
+
+	trig_cnt++;
+
+	/* ism330dhcx accel */
+	sensor_sample_fetch_chan(dev, SENSOR_CHAN_ACCEL_XYZ);
+	sensor_channel_get(dev, SENSOR_CHAN_ACCEL_X, &x);
+	sensor_channel_get(dev, SENSOR_CHAN_ACCEL_Y, &y);
+	sensor_channel_get(dev, SENSOR_CHAN_ACCEL_Z, &z);
+
+	printf("accel x:%f m/s2 y:%f m/s2 z:%f m/s2\n",
+			out_ev(&x), out_ev(&y), out_ev(&z));
+
+	/* ism330dhcx gyro */
+	sensor_sample_fetch_chan(dev, SENSOR_CHAN_GYRO_XYZ);
+	sensor_channel_get(dev, SENSOR_CHAN_GYRO_X, &x);
+	sensor_channel_get(dev, SENSOR_CHAN_GYRO_Y, &y);
+	sensor_channel_get(dev, SENSOR_CHAN_GYRO_Z, &z);
+
+	printf("gyro x:%f dps y:%f dps z:%f dps\n",
+			out_ev(&x), out_ev(&y), out_ev(&z));
+
+	printf("trig_cnt:%d\n\n", trig_cnt);
+}
+
+static int set_sampling_freq(const struct device *dev)
+{
+	int ret = 0;
+	struct sensor_value odr_attr;
+
+	/* set accel/gyro sampling frequency to 12.5 Hz */
+	odr_attr.val1 = 12.5;
```

Comment:
How does this work? val1 is not a floating point variable.

Reply:
Thanks for pointing that out. There is an implicit type casting here. In the ism330dhcx driver code, it only uses the val1 of the struct sensor_value, and it maintains an odr map that maps value of 12 to the odr of 12.5Hz.

I have updated the code in order to make it clear that the actual assigned value is 12, but this value is intended to set the sampling frequency to 12.5 Hz. Hopefully this can help to reduce the confusion.